libvirt/domain_def.go

```diff
@@ -1,216 +1,76 @@
-	var value string
-	err := d.DecodeElement(&value, &start)
-	if err != nil {
-		return err
+func newFilesystemDef() libvirtxml.DomainFilesystem {
+	return libvirtxml.DomainFilesystem{
+		Type:       "mount",  // This is the only type used by qemu/kvm
+		AccessMode: "mapped", // A safe default value
+		ReadOnly:   &libvirtxml.DomainFilesystemReadOnly{},
 	}
-
-	*r = value == ""
-
-	return nil
-}
-
-func newFilesystemDef() defFilesystem {
-	fs := defFilesystem{}
-
-	// This is the only type used by qemu/kvm
-	fs.Type = "mount"
-
-	// A safe default value
-	fs.AccessMode = "mapped"
-	fs.ReadOnly = true
-
-	return fs
 }
 
 // Creates a domain definition with the defaults
 // the provider uses
-func newDomainDef() defDomain {
-	// libvirt domain definition
-	domainDef := defDomain{}
+func newDomainDef() libvirtxml.Domain {
+	domainDef := libvirtxml.Domain{
+		OS: &libvirtxml.DomainOS{
+			Type: &libvirtxml.DomainOSType{
+				Type: "hvm",
+			},
+		},
+		Memory: &libvirtxml.DomainMemory{
+			Unit:  "MiB",
+			Value: 512,
+		},
+		VCPU: &libvirtxml.DomainVCPU{
+			Placement: "static",
+			Value:     1,
+		},
+		CPU:     &libvirtxml.DomainCPU{},
+		Devices: &libvirtxml.DomainDeviceList{},
+		Features: &libvirtxml.DomainFeatureList{
+			PAE:  &libvirtxml.DomainFeature{},
+			ACPI: &libvirtxml.DomainFeature{},
+			APIC: &libvirtxml.DomainFeatureAPIC{},
+		},
+	}
+
 	if v := os.Getenv("TERRAFORM_LIBVIRT_TEST_DOMAIN_TYPE"); v != "" {
 		domainDef.Type = v
 	} else {
 		domainDef.Type = "kvm"
 	}
-	domainDef.Xmlns = ""
-
-	domainDef.Os = defOs{}
-	domainDef.Os.Type = defOsType{}
-	domainDef.Os.Type.Name = "hvm"
 
-	domainDef.Memory = defMemory{}
-	domainDef.Memory.Unit = "MiB"
-	domainDef.Memory.Amount = 512
-
-	domainDef.VCpu = defVCpu{}
-	domainDef.VCpu.Placement = "static"
-	domainDef.VCpu.Amount = 1
-
-	domainDef.Devices.Graphics = &defGraphics{}
-	domainDef.Devices.Graphics.Type = "spice"
-	domainDef.Devices.Graphics.Autoport = "yes"
-	domainDef.Devices.Graphics.Listen.Type = "none"
-
-	domainDef.Devices.QemuGAChannel.Type = "unix"
-	domainDef.Devices.QemuGAChannel.Source.Mode = "bind"
-	domainDef.Devices.QemuGAChannel.Target.Type = "virtio"
-	domainDef.Devices.QemuGAChannel.Target.Name = "org.qemu.guest_agent.0"
+	graphic := libvirtxml.DomainGraphic{
+		Type:     "spice",
+		AutoPort: "yes",
+	}
+	domainDef.Devices.Graphics = append(domainDef.Devices.Graphics, graphic)
+
+	channel := libvirtxml.DomainChannel{
+		Type: "unix",
+		Source: &libvirtxml.DomainChardevSource{
+			Mode: "bind",
+		},
+		Target: &libvirtxml.DomainChannelTarget{
+			Type: "virtio",
+			Name: "org.qemu.guest_agent.0",
+		},
+	}
+	domainDef.Devices.Channels = append(domainDef.Devices.Channels, channel)
 
-	domainDef.Devices.Rng.Model = "virtio"
-	domainDef.Devices.Rng.Backend.Model = "random"
+	rng := libvirtxml.DomainRNG{
+		Model: "virtio",
+		Backend: &libvirtxml.DomainRNGBackend{
+			Model: "random",
+		},
+	}
+	domainDef.Devices.RNGs = append(domainDef.Devices.RNGs, rng)
```

Comment:
Couldn't we integrate these definitions in the previous `domainDef` definition? Something like:

```Go
	domainDef := libvirtxml.Domain{
		OS: &libvirtxml.DomainOS{
			Type: &libvirtxml.DomainOSType{
				Type: "hvm",
			},
		},
		Memory: &libvirtxml.DomainMemory{
			Unit:  "MiB",
			Value: 512,
		},
		VCPU: &libvirtxml.DomainVCPU{
			Placement: "static",
			Value:     1,
		},
		CPU:     &libvirtxml.DomainCPU{},
		Devices: &libvirtxml.DomainDeviceList{
                                libvirtxml.DomainChannel{
                		                Type: "unix",
                                		Source: &libvirtxml.DomainChardevSource{
                                			Mode: "bind",
                		                },
                                		Target: &libvirtxml.DomainChannelTarget{
		                                	Type: "virtio",
                                			Name: "org.qemu.guest_agent.0",
                		                },
                                	}
                },
		Features: &libvirtxml.DomainFeatureList{
			PAE:  &libvirtxml.DomainFeature{},
			ACPI: &libvirtxml.DomainFeature{},
			APIC: &libvirtxml.DomainFeatureAPIC{},
		},
	}
```

(and the same for libvirtxml.DomainRNG)